src/content/en/2020/http2.md

```diff
@@ -4,20 +4,534 @@ part_number: IV
+
+HTTP/2 introduced the concept of prioritization to help the client communicate to the server how it thinks the multiplexing should be done. Every stream is assigned a weight (how much of the available bandwidth the stream should be allocated) and possibly a parent (another stream which should be delivered first). With the flexibility of HTTP/2’s prioritization model, it is not altogether surprising that all of the current browser engines implemented [different prioritization strategies](https://calendar.perfplanet.com/2018/http2-prioritization/), none of which are [optimal](https://www.youtube.com/watch?v=nH4iRpFnf1c).
+
+There are also problems on the server side, leading to many servers either implementing prioritization poorly or not at all.  In the case of HTTP/1.x, tuning the server-side send buffers to be as big as possible has no downside other than the increase in memory use (trading off memory for CPU) and is an effective way to increase the throughput of a web server. This is not true for HTTP/2, as data in the send buffer cannot be re-prioritized if a request for a new, more important request comes in.  For an HTTP/2 server, the optimal send buffer size is thus the minimum amount of data required to fully utilize the available bandwidth. This allows the server to respond immediately if a higher-priority response is received.
+
+This problem of large buffers messing with (re-)prioritization also exists in the network where it goes by the name “bufferbloat”.  Network equipment would rather buffer packets than drop them when there’s a short burst.  However, if the server sends more data than the path to the client can consume, these buffers fill to capacity. These bytes already “stored” on the network limit the server’s ability to send a higher-priority response earlier, just as a large send buffer does. To minimize the amount of data held in buffers, [a recent congestion control algorithm such as BBR should be used](https://blog.cloudflare.com/http-2-prioritization-with-nginx/).
+
+A [test suite](https://github.com/andydavies/http2-prioritization-issues) and a list of how various CDN/Cloud Hosting services perform was created. The bad news is that only 9 of the 36 prioritize correctly. Figure 22.16 shows that for sites using a CDN, around 31.7% do not prioritize correctly. This is up from 26.82% last year, mainly due to the increase in Google CDN usage. Rather than relying on the browser-sent priorities, there are some servers that implement a [server side prioritization](https://blog.cloudflare.com/better-http-2-prioritization-for-a-faster-web/) scheme instead, improving upon the browser’s hints with additional logic.
```

Comment:
```suggestion
This [test suite](https://github.com/andydavies/http2-prioritization-issues) measures and reports how various CDN/Cloud Hosting services perform. The bad news is that only 9 of the 36 prioritize correctly. Figure 22.16 shows that for sites using a CDN, around 31.7% do not prioritize correctly. This is up from 26.82% last year, mainly due to the increase in Google CDN usage. Rather than relying on the browser-sent priorities, there are some servers that implement a [server side prioritization](https://blog.cloudflare.com/better-http-2-prioritization-for-a-faster-web/) scheme instead, improving upon the browser’s hints with additional logic.
```